poh/src/leader_bank_notifier.rs

```diff
@@ -0,0 +1,289 @@
+    slot: Option<Slot>,
+    bank: Weak<Bank>,
+}
+
+impl LeaderBankNotifier {
+    /// Set the status to `InProgress` and notify any waiting threads.
+    /// Panics if the status is not `StandBy` - cannot have multiple
+    /// leader banks in progress.
+    pub fn set_in_progress(&self, bank: &Arc<Bank>) {
+        let mut state = self.state.lock().unwrap();
+        assert_eq!(state.status, Status::StandBy);
+
+        *state = SlotAndBankWithStatus {
+            status: Status::InProgress,
+            slot: Some(bank.slot()),
+            bank: Arc::downgrade(bank),
+        };
+        drop(state);
+
+        self.condvar.notify_all();
+    }
+
+    /// Set the status to `StandBy` and notify any waiting threads if the status was changed.
+    /// Panics if the stored slot does not match the given slot.
+    pub fn set_completed(&self, slot: Slot) {
+        let mut state = self.state.lock().unwrap();
+        assert_eq!(state.slot, Some(slot));
+
+        // `set_completed` can be called multiple times for the same slot because it
+        // may be called from multiple threads.
+        if matches!(state.status, Status::StandBy) {
+            return;
+        }
+
+        state.status = Status::StandBy;
+        drop(state);
+
+        self.condvar.notify_all();
+    }
+
+    /// If the status is `InProgress`, immediately return a weak reference to the bank.
+    /// Otherwise, wait up to the `timeout` for the status to become `InProgress`.
+    /// If the timeout is reached, the weak reference is unupgradable.
+    pub fn get_or_wait_for_in_progress(&self, timeout: Duration) -> Weak<Bank> {
+        let state = self.state.lock().unwrap();
+        self.get_or_wait_for_in_progress_state(timeout, state)
+            .map(|state| state.bank.clone())
+            .unwrap_or_else(Weak::new)
+    }
+
+    /// Wait for next notification for a completed leader slot.
+    /// Returns `None` if the timeout is reached
+    pub fn wait_for_completed(&self, mut remaining_timeout: Duration) -> Option<Slot> {
+        let state = self.state.lock().unwrap();
+
+        // If currently `StandBy`, need to wait for `InProgress` to begin.
+        let now = Instant::now();
+        let state = self.get_or_wait_for_in_progress_state(remaining_timeout, state);
+
+        let Some(state) = state else { return None; };
+        remaining_timeout = remaining_timeout.saturating_sub(now.elapsed());
+
+        // Wait for `StandBy` to be set.
+        let (state, wait_timeout_result) = self
+            .condvar
+            .wait_timeout_while(state, remaining_timeout, |state| {
+                matches!(state.status, Status::InProgress)
+            })
+            .unwrap();
+
+        if !wait_timeout_result.timed_out() {
+            state.slot
+        } else {
+            None
+        }
+    }
+
+    // Helper function to get or wait for the `InProgress` status with a given `MutexGuard`.
+    fn get_or_wait_for_in_progress_state<'a>(
+        &self,
```

Comment:
nit: if this `fn` is called _a helper function_, I'd like not to take `&self`, rather `&'a Condvar`, so that it's very clear the intent to readers and avoids pitfalls like double locking. (admittedly, code is small, so pretty low risk).

Alternatively, if this `fn` is called _a helper method_, `&self` is ok-ish. but, I'd prefer adjusting the code.